### DISCUSSION TITLE
Trying to build on windows with mingw64

### DISCUSSION THREAD

**Discussion by dsacks** [Clozure/ccl#449] [General]
I'm on windows 10 trying to build with msys/mingw64 tools. 

doing make in ccl/lisp-kernel/win64, I'm getting a bunch  of errors including:
thread_manager.c:242: undefined reference to `atomic_swap'
thread_manager.c:602: undefined reference to `atomic_ior'

etc. I assume I'm missing something, but I don't know what.

**Comment by dsacks** [comment]
Actually, I'm missing the right way to build CCL. I looked at the build instructions and am on the right track now.  I tried building the latest from git with my working CCL build, but am getting an error indicating I'm missing things from my mingwin installation;

I don't know why, but I have ld, but not x86_64-w64-mingw32-ld and it looks like I don't have m4, but that is probably a pacman command away. I've yet to figure out what is up with x86_64-w64-mingw32-ld, but in any case those are not CCL issues.

**Reply by dsacks** [reply]
Okay, silly question, but I think I succeeded in building the latest CCL. However, how do I execute it? These are the last few lines of the output from (rebuild-ccl :clean t)
;Loading #P"C:/Users/dennis/scoop/apps/ccl/current/level-0/nfasload.wx64fsl"...
;Wrote bootstrapping image: #P"C:/Users/dennis/scoop/apps/ccl/current/wx86-boot64.image"
;Wrote heap image: #P"C:/Users/dennis/scoop/apps/ccl/current/wx86cl64.image"
NIL

**Comment by xrme** [comment]
There are two steps.

One is to build the lisp kernel. On Windows, you'd go into `lisp-kernel/win64` and then run `make`. This will create an executable named `wx86cl64.exe`.

The next step is to rebuild the heap image file. From the top of the ccl directory, start your ccl with `./wx86cl64` and then evaluate `(rebuild-ccl :clean t)`. It looks that that's what you just did.

When you've performed those two steps, you're done.

**Reply by dsacks** [reply]
Do you compile using msys? I tried doing a make clean and make under lisp-kernel/win64 and when I try to execute the resulting binary it gives me this error:

bash: ./wx86cl64.exe: cannot execute binary file: Exec format error

$ file ../../wx86cl64.exe
../../wx86cl64.exe: PE32+ executable (console) x86-64, for MS Windows, 23 sections

the last part of the make output is weird to me:
/mingw64/bin/x86_64-w64-mingw32-gcc -Wl,--image-base=0x10000 -Wl,-script=pei-x86-64.x -m64 -g    -o
../../wx86cl64.exe  pad.o x86-spjump64.o x86-spentry64.o x86-subprims64.o pmcl-kernel.o gc-common.o
x86-gc.o bits.o  x86-exceptions.o x86-utils.o image.o thread_manager.o lisp-debug.o memory.o windows
-calls.o x86-asmutils64.o  imports.o lispdcmd.o plprint.o plsym.o xlbt.o x86_print.o -lpsapi -lws2_3
2 -static -lpthread
**C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe
: ../../wx86cl64.exe:/4: section below image base
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe
: ../../wx86cl64.exe:/20: section below image base
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe
: ../../wx86cl64.exe:/36: section below image base**

**Comment by xrme** [comment]
I just checked in changes so that the Windows lisp kernels can be built with an up-to-date MSYS2 installation. See #425, and the comments in win64/Makefile.

If you'll pull and try again (use the MINGW64 environment), you should be able to build a working `wx86cl64.exe`.
